x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/aggregation/blockhash/BytesRef3BlockHash.java

```diff
@@ -0,0 +1,220 @@
+import org.elasticsearch.compute.data.Block;
+import org.elasticsearch.compute.data.BlockFactory;
+import org.elasticsearch.compute.data.BytesRefBlock;
+import org.elasticsearch.compute.data.BytesRefVector;
+import org.elasticsearch.compute.data.IntBlock;
+import org.elasticsearch.compute.data.IntVector;
+import org.elasticsearch.compute.data.Page;
+import org.elasticsearch.core.ReleasableIterator;
+import org.elasticsearch.core.Releasables;
+
+import java.util.Locale;
+
+/**
+ * Maps three {@link BytesRefBlock}s to group ids.
+ */
+final class BytesRef3BlockHash extends BlockHash {
+    private final int emitBatchSize;
+    private final int channel1;
+    private final int channel2;
+    private final int channel3;
+    private final BytesRefBlockHash hash1;
+    private final BytesRefBlockHash hash2;
+    private final BytesRefBlockHash hash3;
+    private final Int3Hash finalHash;
+
+    BytesRef3BlockHash(BlockFactory blockFactory, int channel1, int channel2, int channel3, int emitBatchSize) {
+        super(blockFactory);
+        this.emitBatchSize = emitBatchSize;
+        this.channel1 = channel1;
+        this.channel2 = channel2;
+        this.channel3 = channel3;
+        boolean success = false;
+        try {
+            this.hash1 = new BytesRefBlockHash(channel1, blockFactory);
+            this.hash2 = new BytesRefBlockHash(channel2, blockFactory);
+            this.hash3 = new BytesRefBlockHash(channel3, blockFactory);
+            this.finalHash = new Int3Hash(1, blockFactory.bigArrays());
+            success = true;
+        } finally {
+            if (success == false) {
+                close();
+            }
+        }
+    }
+
+    @Override
+    public void close() {
+        Releasables.close(hash1, hash2, hash3, finalHash);
+    }
+
+    @Override
+    public void add(Page page, GroupingAggregatorFunction.AddInput addInput) {
+        BytesRefBlock b1 = page.getBlock(channel1);
+        BytesRefBlock b2 = page.getBlock(channel2);
+        BytesRefBlock b3 = page.getBlock(channel3);
+        BytesRefVector v1 = b1.asVector();
+        BytesRefVector v2 = b2.asVector();
+        BytesRefVector v3 = b3.asVector();
+        if (v1 != null && v2 != null && v3 != null) {
+            addVectors(v1, v2, v3, addInput);
+        } else {
+            try (IntBlock k1 = hash1.add(b1); IntBlock k2 = hash2.add(b2); IntBlock k3 = hash3.add(b3)) {
+                try (AddWork work = new AddWork(k1, k2, k3, addInput)) {
+                    work.add();
+                }
+            }
+        }
+    }
+
+    private void addVectors(BytesRefVector v1, BytesRefVector v2, BytesRefVector v3, GroupingAggregatorFunction.AddInput addInput) {
+        final int positionCount = v1.getPositionCount();
+        try (IntVector.Builder ordsBuilder = blockFactory.newIntVectorFixedBuilder(positionCount)) {
+            // TODO: enable ordinal vectors in BytesRefBlockHash
```

Comment:
One day I'd like to do away with the special ordinals aggregator and use ordinals here. But that's a problem for another day.

Reply:
+1